src/main/java/org/candlepin/resource/OwnerProductResource.java

```diff
@@ -24,302 +24,706 @@
+    @Transactional
+    public ProductDTO getProductById(@Verify(Owner.class) String ownerKey, String productId) {
         Owner owner = this.getOwnerByKey(ownerKey);
-        Product product = this.fetchProduct(owner, productId);
+        Product product = this.resolveProductId(owner, productId, null);
 
         return this.translator.translate(product, ProductDTO.class);
     }
 
     @Override
     @Transactional
-    public ProductCertificateDTO getProductCertificateByOwner(String ownerKey, String productId) {
+    public ProductCertificateDTO getProductCertificateById(String ownerKey, String productId) {
         if (!productId.matches("\\d+")) {
             throw new BadRequestException(i18n.tr("Only numeric product IDs are allowed."));
         }
 
         Owner owner = this.getOwnerByKey(ownerKey);
-        Product product = this.fetchProduct(owner, productId);
+        Product product = this.resolveProductId(owner, productId, null);
 
+        // FIXME: Should this be fetching the cert from the product service? Feels like an oversight.
         ProductCertificate productCertificate = this.productCertCurator.getCertForProduct(product);
         return this.translator.translate(productCertificate, ProductCertificateDTO.class);
     }
 
     @Override
-    public CandlepinQuery<ProductDTO> getProductsByOwner(@Verify(Owner.class) String ownerKey,
-        List<String> productIds) {
+    @Transactional
+    public Stream<ProductDTO> getProductsByOwner(@Verify(Owner.class) String ownerKey,
+        List<String> productIds, Boolean omitGlobalEntities) {
 
-        Owner owner = getOwnerByKey(ownerKey);
-        CandlepinQuery<Product> query = productIds != null && !productIds.isEmpty() ?
-            this.ownerProductCurator.getProductsByIds(owner, productIds) :
-            this.ownerProductCurator.getProductsByOwnerCPQ(owner);
+        Owner owner = this.getOwnerByKey(ownerKey);
+        String namespace = owner.getKey();
+
+        Collection<Product> products;
+
+        if (omitGlobalEntities == null || !omitGlobalEntities) {
+            products = productIds != null && !productIds.isEmpty() ?
+                this.productCurator.resolveProductIds(namespace, productIds).values() :
+                this.productCurator.resolveProductsByNamespace(namespace);
+        }
+        else {
+            products = productIds != null && !productIds.isEmpty() ?
+                this.productCurator.getProductsByIds(namespace, productIds).values() :
+                this.productCurator.getProductsByNamespace(namespace);
+        }
+
+        Stream<ProductDTO> stream = products.stream()
+            .map(this.translator.getStreamMapper(Product.class, ProductDTO.class));
 
-        return this.translator.translateQuery(query, ProductDTO.class);
+        return this.pagingUtilFactory.forClass(ProductDTO.class)
+            .applyPaging(stream, products.size());
```

Comment:
I didnt have a chance to fully look through the paging utility PR, but do I understand correctly that it applies paging after the fact/db query? So the db query may return say, 10000 results, all of those are fetched into memory, and then this util returns only the page requested by the client, so e.g. items 201-300?
If so, isn't this problematic memory-wise for endpoints/queries that fetch a lot of results, and paging at the query level was supposed to be a solution here?

Reply:
Also, this is not really a problem for this specific endpoint, since based on past investigation I see will not return enough products to DoS candlepin, but for a few other endpoints it will be

Reply:
Yes, your understanding of how the pagingutil works is correct; and it is not meant to be a replacement for properly mapping and narrowing queries in areas where we've determined the output needs to be limited (consumer fetching, for instance).

Given where the JPA interfaces are combined with our current timelines, I don't foresee us having the time to spec out some magic that can (a) translate and validate user-facing sortBy fields to internal model field names, (b) apply the query limitations generally in a way that we wouldn't need to hand-craft individual queries, and (c) get the total element count as a separate query, again, without needing to hand-craft it for the operation at hand.

The interface here is general enough that we can use it to apply paging quickly in places where the base query is safe enough that we allow it to be used without paging (of which, products and content should be), but not in heavier places where our fetch-all query is going to return 6+ figures of results.